### PR TITLE
[MM-13746] Add GetTeamMembersForUser and GetChannelMembersForUser apis

### PR DESCRIPTION
**Pull request by gupsho**
Add GetTeamMembersForUser and GetChannelMembersForUser apis

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/10096

#### Checklist
- [x] Added or updated unit tests (required for all new features)

### CODE REVIEW

File: plugin/api.go [hunk 1]
```diff
@@ -174,6 +174,9 @@ type API interface {
 	// GetTeamMember returns a specific membership.
 	GetTeamMember(teamId, userId string) (*model.TeamMember, *model.AppError)
 
+	// GetTeamMembersForUser returns team members for all teams a user is on
```
Comment (hanzei) [inline plugin/api.go]: Can you please document the minimum server version here?

```suggestion
	// GetTeamMembersForUser returns all team memberships for a user.
	// Minimum server version: 5.6
```

File: plugin/api.go [hunk 2]
```diff
@@ -242,6 +245,9 @@ type API interface {
 	// Minimum server version: 5.6
 	GetChannelMembersByIds(channelId string, userIds []string) (*model.ChannelMembers, *model.AppError)
 
+	// GetChannelMembersForUser gets channel membership for all users on a team that a user belongs
```
Comment (hanzei) [inline plugin/api.go]: Same here

```suggestion
	// GetChannelMembersForUser returns all channel memberships on a team for a user.
	// Minimum server version: 5.6
```

File: store/sqlstore/channel_store.go [hunk 3]
```diff
@@ -796,10 +796,10 @@ func (s SqlChannelStore) SetDeleteAt(channelId string, deleteAt, updateAt int64)
 		// Additionally propagate the write to the PublicChannels table.
 		if _, err := transaction.Exec(`
 			UPDATE
-			    PublicChannels 
-			SET 
+			    PublicChannels
```
Comment (hanzei) [inline store/sqlstore/channel_store.go]: Can you please remove this unrelated changes? It would be appreciated, if you open a separate PR for this.

File: store/storetest/channel_store.go [hunk 4]
```diff
@@ -1419,6 +1420,49 @@ func testChannelStoreGetMembersForUser(t *testing.T, ss store.Store) {
...
+	m2.ChannelId = o2.Id
+	m2.UserId = m1.UserId
+	m2.NotifyProps = model.GetDefaultChannelNotifyProps()
+	store.Must(ss.Channel().SaveMember(&m2))
+
+	cresult := <-ss.Channel().GetMembersForUserWithPagination(o1.TeamId, m1.UserId, 1, 0)
+	members := cresult.Data.(*model.ChannelMembers)
+
+	// no unread messages
+	if len(*members) != 1 {
```
Comment (hanzei) [inline store/storetest/channel_store.go]: Nit: Can you please use testify here:

```suggestion
	assert.Len(t, *members, 1)
```

File: store/storetest/channel_store.go [hunk 5]
```diff
@@ -1419,6 +1420,49 @@ func testChannelStoreGetMembersForUser(t *testing.T, ss store.Store) {
...
+	m1.NotifyProps = model.GetDefaultChannelNotifyProps()
+	store.Must(ss.Channel().SaveMember(&m1))
+
+	m2 := model.ChannelMember{}
+	m2.ChannelId = o2.Id
+	m2.UserId = m1.UserId
+	m2.NotifyProps = model.GetDefaultChannelNotifyProps()
+	store.Must(ss.Channel().SaveMember(&m2))
+
+	cresult := <-ss.Channel().GetMembersForUserWithPagination(o1.TeamId, m1.UserId, 1, 0)
```
Comment (hanzei) [inline store/storetest/channel_store.go]: Maybe I'm mistaken, but should this be 

```suggestion
	cresult := <-ss.Channel().GetMembersForUserWithPagination(o1.TeamId, m1.UserId, 0, 1)
```
?

File: store/storetest/channel_store.go [hunk 6]
```diff
@@ -1419,6 +1420,49 @@ func testChannelStoreGetMembersForUser(t *testing.T, ss store.Store) {
...
+	members := cresult.Data.(*model.ChannelMembers)
+
+	// no unread messages
+	if len(*members) != 1 {
+		t.Fatal("wrong number of members")
+	}
```
Comment (hanzei) [inline store/storetest/channel_store.go]: Would you mind adding a second test for the 2nd membership?

File: store/storetest/team_store.go [hunk 7]
```diff
@@ -668,6 +669,100 @@ func testTeamMembers(t *testing.T, ss store.Store) {
...
 
+func testTeamMembersWithPagination(t *testing.T, ss store.Store) {
+	teamId1 := model.NewId()
+	teamId2 := model.NewId()
+
+	m1 := &model.TeamMember{TeamId: teamId1, UserId: model.NewId()}
+	m2 := &model.TeamMember{TeamId: teamId1, UserId: model.NewId()}
+	m3 := &model.TeamMember{TeamId: teamId2, UserId: model.NewId()}
+
+	if r1 := <-ss.Team().SaveMember(m1, -1); r1.Err != nil {
```
Comment (hanzei) [inline store/storetest/team_store.go]: As stated above: Please use testify

```suggestion
	r1 := <-ss.Team().SaveMember(m1, -1)
	require.Nil(t, r1.Err)
```

File: store/storetest/team_store.go [hunk 8]
```diff
@@ -668,6 +669,100 @@ func testTeamMembers(t *testing.T, ss store.Store) {
...
+		t.Fatal(r1.Err)
+	}
+
+	store.Must(ss.Team().SaveMember(m2, -1))
+	store.Must(ss.Team().SaveMember(m3, -1))
+
+	if r1 := <-ss.Team().GetMembers(teamId1, 0, 100); r1.Err != nil {
```
Comment (hanzei) [inline store/storetest/team_store.go]: Same here:

```suggestion
	r1 := <-ss.Team().GetMembers(teamId1, 0, 100)
	require.Nil(t, r1.Err)
```

**Comment by hanzei** [comment]
Looks like your tests are failing: https://build.mattermost.com/blue/organizations/jenkins/mp%2Fmattermost-server-pr-new/detail/PR-10269/3/tests

File: store/storetest/channel_store.go (hunk 5 above)
Comment (gupsho) [reply store/storetest/channel_store.go]: I was trying to get 1 member back when there are two. First param is perPage and second is pageNum, so it should be 1, 0.

**Comment by gupsho** [comment]
> Looks like your tests are failing: https://build.mattermost.com/blue/organizations/jenkins/mp%2Fmattermost-server-pr-new/detail/PR-10269/3/tests

Hi @hanzei, could you please tell me how can I run just store tests? I am not able to figure out why tests are failing and cannot run a single test.

I have tried these two ways.

 go test -v -run='TestChannelStore' ./store/storetest
?   	github.com/mattermost/mattermost-server/store/storetest	[no test files]
go test -v -run='TestChannelStore' ./store
?   	github.com/mattermost/mattermost-server/store	[no test files]

**Comment by hanzei** [comment]
Hi @gupsho,

You can run these test with `$ go test ./store/sqlstore/ -v -run "TestChannelStore/MySQL/GetMembersForUserWithPagination"` and `$ go test ./store/sqlstore/ -v -run "TestTeamStore/MySQL/GetTeamsForUserWithPagination"`. I know, it a bit tricky. :wink:

**Comment by gupsho** [comment]
Fixed the tests.

File: store/storetest/channel_store.go (hunk 5 above)
Comment (gupsho) [reply store/storetest/channel_store.go]: Fixed.

File: plugin/api.go [hunk 9]
```diff
@@ -239,14 +244,19 @@ type API interface {
 
 	// GetChannelMembers gets a channel membership for all users.
 	//
-	// Minimum server version: 5.6
+	// Minimum server version: 5.8
```
Comment (hanzei) [inline plugin/api.go]: Why did this got changed?

Comment (gupsho) [reply plugin/api.go]: Was a mistake. Fixed.

**Comment by gupsho** [comment]
> One last question. The other parts of the code look good.

Was a mistake. Fixed that comment.

File: plugin/api.go [hunk 10]
```diff
@@ -247,6 +252,11 @@ type API interface {
 	// Minimum server version: 5.6
 	GetChannelMembersByIds(channelId string, userIds []string) (*model.ChannelMembers, *model.AppError)
 
+	// GetChannelMembersForUser returns all channel memberships on a team for a user.
+	//
+	// Minimum server version: 5.6
+	GetChannelMembersForUser(teamId, userId string, page, perPage int) (*model.ChannelMembers, *model.AppError)
```
[the remark this answers is not in the document]
Comment (hanzei) [reply plugin/api.go]: This is also the way proposed in the original ticket. I'm 2/5 on replacing `*model.ChannelMembers` with `[]*model.ChannelMember`.

[the remark this answers is not in the document]
Comment (gupsho) [reply plugin/api.go]: Sure, let me change that.

[the remark this answers is not in the document]
Comment (gupsho) [reply plugin/api.go]: Hey @hanzei ,

Fixed for the new changes. Do you want me to change for all the apis which return *model.ChannelMembers?

[the remark this answers is not in the document]
Comment (hanzei) [reply plugin/api.go]: Thanks for fixing this.

[the remark this answers is not in the document]
Comment (hanzei) [reply plugin/api.go]: Your last changes arn't quite correct. I created a plugin that call the two new methods. But when calling them, the server panics with:
```
panic: interface conversion: interface {} is *model.ChannelMembers, not []*model.ChannelMember

goroutine 4914 [running]:
github.com/mattermost/mattermost-server/app.(*App).GetChannelMembersForUserWithPagination(0xc001754540, 0xc004277640, 0x1a, 0xc004277660, 0x1a, 0x0, 0x64, 0x12a4cc0, 0x14300a0, 0x40d801, ...)
	/home/$USER/go/src/github.com/mattermost/mattermost-server/app/channel.go:1215 +0x15e
github.com/mattermost/mattermost-server/app.(*PluginAPI).GetChannelMembersForUser(0xc001a4d1d0, 0xc004277640, 0x1a, 0xc004277660, 0x1a, 0x0, 0x64, 0xc000123880, 0x18, 0x7f0c74c46440,
	/home/$USER/go/src/github.com/mattermost/mattermost-server/app/plugin_api.go:383 +0x7b
github.com/mattermost/mattermost-server/plugin.(*apiRPCServer).GetChannelMembersForUser(0xc001c45060, 0xc003a30cc0, 0xc001f26aa0, 0x0, 0x0)
	/home/$USER/go/src/github.com/mattermost/mattermost-server/plugin/client_rpc_generated.go:2263 +0xa3
reflect.Value.call(0xc001791200, 0xc00248f668, 0x13, 0x1443088, 0x4, 0xc001e3df18, 0x3, 0x3, 0xc002545100, 0x4070bc, ...)
	/home/$USER/.local/go/src/reflect/value.go:447 +0x454
reflect.Value.Call(0xc001791200, 0xc00248f668, 0x13, 0xc001a58f18, 0x3, 0x3, 0xc001a58f38, 0x3, 0x0)
	/home/$USER/.local/go/src/reflect/value.go:308 +0xa4
net/rpc.(*service).call(0xc0002485c0, 0xc001cd2280, 0xc0020021c8, 0xc0020021e0, 0xc001b9fc00, 0xc0013dcec0, 0x11d74e0, 0xc003a30cc0, 0x16, 0x11d7520, ...)
	/home/$USER/.local/go/src/net/rpc/server.go:384 +0x14e
created by net/rpc.(*Server).ServeCodec
	/home/$USER/.local/go/src/net/rpc/server.go:481 +0x47e
exit status 2          
```

**Comment by hanzei** [comment]
Since the methods have become a bit more complex and tricky (like you saw with the panic above), it would be great if you would add two tests to `app/plugin_api_test.go`. I can take care of this, if you don't have the time.

File: plugin/api.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (levb) [reply plugin/api.go]: @hanzei may or may not be your issue, but when building plugins that rely on a recent (local) server change I often forget to un-vendor mattermost-server from the plugin build.

I'd LGTM now, but I'd like to see the ^^ issue resolved first.

OT, 1/5 I wonder if we should add "real plugin RPC" tests for new plugin APIs, similar to https://github.com/mattermost/mattermost-server/blob/master/app/plugin_hooks_test.go

**Comment by gupsho** [comment]
> Since the methods have become a bit more complex and tricky (like you saw with the panic above), it would be great if you would add two tests to `app/plugin_api_test.go`. I can take care of this, if you don't have the time.

Let me do that. Apologies for the trouble. Will have it up for review later today.

File: plugin/api.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply plugin/api.go]: > may or may not be your issue, but when building plugins that rely on a recent (local) server change I often forget to un-vendor mattermost-server from the plugin build.

Thanks for the hint. But I double checked and this is not the cause of the problem.

**Comment by gupsho** [comment]
> > Since the methods have become a bit more complex and tricky (like you saw with the panic above), it would be great if you would add two tests to `app/plugin_api_test.go`. I can take care of this, if you don't have the time.
> 
> Let me do that. Apologies for the trouble. Will have it up for review later today.

Hopefully should work now. Added new tests and fixed the panic.

File: plugin/api.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (gupsho) [reply plugin/api.go]: Fixed the panic. Should work now.

File: app/plugin_api_test.go [hunk 11]
```diff
@@ -881,3 +881,31 @@ func TestPluginAPI_SearchTeams(t *testing.T) {
...
+func TestPluginAPI_GetTeamMembersForUser(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+	api := th.SetupPluginAPI()
+
+	userId := th.BasicUser.Id
+	teamMembers, err := api.GetTeamMembersForUser(userId, 0, 10)
+	assert.Nil(t, err)
+	assert.Equal(t, len(teamMembers), 1)
+	fmt.Println(teamMembers[0])
```
Comment (hanzei) [inline app/plugin_api_test.go]: You forgot this one :wink:

Comment (gupsho) [reply app/plugin_api_test.go]: Fixed.
